src/common/dao/blob.go

```diff
@@ -64,6 +65,107 @@ func DeleteBlob(digest string) error {
 	return err
 }
 
+// ListBlobs list blobs according to the query conditions
+func ListBlobs(query *models.BlobQuery) ([]*models.Blob, error) {
+	qs := GetOrmer().QueryTable(&models.Blob{})
+
+	if query != nil {
+		if query.Digest != "" {
+			qs = qs.Filter("Digest", query.Digest)
+		}
+
+		if query.ContentType != "" {
+			qs = qs.Filter("ContentType", query.ContentType)
+		}
+
+		if len(query.Digests) > 0 {
+			qs = qs.Filter("Digest__in", query.Digests)
+		}
+
+		if query.Size > 0 {
+			qs = qs.Limit(query.Size)
+			if query.Page > 0 {
+				qs = qs.Offset((query.Page - 1) * query.Size)
+			}
+		}
+	}
+
+	blobs := []*models.Blob{}
+	_, err := qs.All(&blobs)
+	return blobs, err
+}
+
+// SyncBlobs sync references to blobs
+func SyncBlobs(references []distribution.Descriptor) error {
+	if len(references) == 0 {
+		return nil
+	}
+
+	var digests []string
+	for _, reference := range references {
+		digests = append(digests, reference.Digest.String())
+	}
+
+	existing, err := ListBlobs(&models.BlobQuery{Digests: digests})
+	if err != nil {
+		return err
+	}
+
+	mp := make(map[string]*models.Blob, len(existing))
+	for _, blob := range existing {
+		mp[blob.Digest] = blob
+	}
+
+	var missing, updating []*models.Blob
+	for _, reference := range references {
+		if blob, found := mp[reference.Digest.String()]; found {
+			if blob.ContentType == "" {
+				blob.ContentType = reference.MediaType
+				updating = append(updating, blob)
+			}
+
+		} else {
+			missing = append(missing, &models.Blob{
+				Digest:       reference.Digest.String(),
+				ContentType:  reference.MediaType,
+				Size:         reference.Size,
+				CreationTime: time.Now(),
+			})
+		}
+	}
+
+	o := GetOrmer()
+
+	if len(updating) > 0 {
+		for _, blob := range updating {
+			if _, err := o.Update(blob, "content_type"); err != nil {
+				log.Warningf("Failed to update blob %s, error: %v", blob.Digest, err)
+			}
+		}
+	}
+
+	if len(missing) > 0 {
+		_, err = o.InsertMulti(10, missing)
+		if err != nil {
+			if strings.Contains(err.Error(), "duplicate key value violates unique constraint") {
+				return ErrDupRows
+			}
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+// CleanupBlobs remove blobs which are not referenced by artifact blob
+func CleanupBlobs() error {
```

Comment:
The IN could in an empty condition? that could lead to a PG error.

Reply:
this sql will work well even artifact_blob is empty, but this func removed for the job of gc will not call it.